skfem/assembly/basis/subdomain_facet_basis.py

```diff
@@ -0,0 +1,80 @@
+import logging
+from typing import Optional, Tuple, List
+
+import numpy as np
+from numpy import ndarray
+from skfem.element import Element, ElementTriP0
+from skfem.mapping import Mapping
+from skfem.mesh import Mesh
+
+from .. import asm
+from .cell_basis import CellBasis
+from .facet_basis import FacetBasis
+from .interior_facet_basis import InteriorFacetBasis
+from ..dofs import Dofs
+from ..form import Functional
+
+logger = logging.getLogger(__name__)
+
+
+class SubdomainFacetBasis(FacetBasis):
```

Comment:
This is the relevant part I presume.

Reply:
> Why the refactoring and so many unrelated changes though? 

Yeah, I just dropped a grenade in, partly by accident. The problems started with me being irked that things things weren't really `BoundaryFacetBasis` and so I thought I could maybe refactor to `FacetBasis` easily enough, and it would all just be okay. So I tried it.

Last time I tried a pull request I was overrun with formatting bugs and someone suggested black. I hadn't used it prior, but now admit its a nice tool... anyway I thought it was the secret sauce to fixing format/style issues. I was wrong. That caused a ton of line changes that I didn't mean, but I didn't realize until the tests started failing that black was perhaps not the answer. By that point it was too late, and you see the end result. :(

> This is the relevant part I presume.

Yes, definitely. And I already feel like it is a train wreck of an implementation. But I hope you see some interesting ideas you hadn't thought of before.

Reply:
Black, which has been discussed a couple of times in scikit-fem, is O. K. to run if it has been adopted across the project, but it hasn't.  I do still use it here, but in a restricted way, for new code, like so:
* Write my additions without any regard for format.
* Run `black` on the modified file.
* Copy my addition from the blackened file.
* Undo all the edits; e.g. `git checkout $THE_FILE`
* Paste my blackened addition into the otherwise original file.